And still more parse tree equivalence understandings

mr-martian opened the pull request:
Perhaps I should have done more of these edits *before* making the pull requests. :)
This time I have added ```===```, ```!==```, ```!=```, ```==```, ```&```, ```|```, ```^```, ```&&```, ```||``` and I have now exhausted the list of JS operators that can be treated this way. They all have tests, of course.

Now for the more interesting part:
```var a = 3``` will now be transformed into ```var a; a = 3``` but not vice versa.
This works, but there are 2 tests that fail:
* [Complex vars with small mismatch breaks.](https://github.com/mr-martian/structuredjs/blob/master/tests.js#L959)
* [Upward expression ordering fails.](https://github.com/mr-martian/structuredjs/blob/master/tests.js#L930)

The former appears to be testing something that is only slightly wrong, and broken by the modification.

The latter is checking that ```while(true) { var x = 5; } var y = 6;``` does not match ```var x = 5; var y = 6;```, which appearantly it now does.

So should I remove this part and just leave every commutative operator? For the moment I have added an option to prevent this change from being applied, which is passed by the tests that failed. This does not seem like an optimal solution, perhaps those particular tests need to be changed?

pamelafox commented:
Hm. Yeah, I figured the var a = 3 transformation would be tricky. I'd say remove that part for now. Having all the operators will be a huge help.

mr-martian commented:
I have commented out the ```var a = 3``` and ```{notvar: true}``` parts. All tests pass.

pamelafox commented on structured.js:
```diff
@@ -585,7 +599,34 @@
                             value: 1,
                             raw: "1"}};
-        }
+        } /*else if ("body" in currTree && !options.notvar) {
```
Can you put a comment about why we chose to comment it out? (This would make bla and bla equivalent, but then while(true) { var x = 5; } var y = 6; will match var x = 5; var y = 6;, which we don't think we want.)
Alternatively, just delete the commented out code. We can retrieve it in the PR history if desired.

pamelafox commented on tests.js:
```diff
@@ -927,7 +927,7 @@ var wildcardVarTests = function() {
         }   \n \
         ";
-        equal(Structured.match(code, structure),
+        equal(Structured.match(code, structure), //, {notvar: true}),
```
Why the commented out {notvar: true} here?

pamelafox commented on tests.js:
```diff
@@ -1903,6 +1903,79 @@ var commutativity = function(){
+        code = "7 || a;"; 
+        equal(!!Structured.match(code, structure, {editorCallbacks: {}}), true, "commutative property of ||");
+
+        /*structure = function() {
```
Please either comment or delete this bit as well. Thanks!

mr-martian commented:
I have removed the commented sections of tests.js and added an explanatory comment on the section in structured.js. I will continue to fiddle with this and see if I can make it work correctly.

pamelafox commented:
Thanks!
